File: OOPforUnfinishedLab3/src/OOPforUnfinishedLab3/MainSystem.java
package OOPforUnfinishedLab3;

import java.util.*;

class Item { //Where i used the concept of objects and classes
    private String pku;
    private String name;
    private String price;
    
    public Item(String pku, String name, String price) {
        this.pku = pku;
        this.name = name;
        this.price = price;
    }
    
    public String getPKU() {
        return pku;
    }
    public String getItemName() {
        return name;
    }
    public String getPrice() {
        return price;
    }
    
    @Override
    public String toString() {
        return "\nPKU: " + pku + "\nName: " + name + "\nPrice: " + price;
    }
}

class Inventory { //Inventory Methods for Employees
    public List<Item> inventory;
    
    public Inventory() {
        this.inventory = new ArrayList<>();
    }
    
    public void addItem() {
        Scanner input = new Scanner(System.in);
        System.out.print("Enter the PKU of item: ");
        String pku = input.nextLine();
        System.out.print("Enter the name of item: ");
        String name = input.nextLine();
        System.out.print("Enter the price of item: ");
        String price = input.nextLine();
        
        Item newItem = new Item(pku, name, price);
        inventory.add(newItem);
        System.out.println("\nNew Item Added:\n" + newItem);
    }
    
    public void deleteItem() {
        if (inventory.isEmpty()) {
            System.out.println("The inventory is already empty.");
        } else {
            Scanner input = new Scanner(System.in);
            System.out.print("Enter the PKU of the item to delete: ");
            String pku = input.nextLine();

            boolean itemFound = false;
            for (int i = 0; i < inventory.size(); i++) {
                if (inventory.get(i).getPKU().equals(pku)) {
                    inventory.remove(i);
                    System.out.println("The book has been deleted successfully.");
                    itemFound = true;
                    break;
                }
            }
            if (!itemFound) {
                System.out.println("The book does not exist.");
            }
        }
    }
    public void findItem(){
        Scanner input = new Scanner(System.in);
        System.out.print("Enter the name of the item: ");
        String name = input.nextLine();

        boolean itemFound = false;
        for (int i = 0; i < inventory.size(); i++) {
            if (inventory.get(i).getItemName().equalsIgnoreCase(name)) {
                System.out.println("Item " + (i) + ":");
                System.out.println(inventory.get(i));
            }
        }
    }
    
    public void viewInventory() {
        if (inventory.isEmpty()) {
            System.out.println("The inventory is empty.");
        } else {
            System.out.println("Inventory Catalog:");
            for (int i = 0; i < inventory.size(); i++) {
                System.out.println("Item " + (i + 1) + ":");
                System.out.println(inventory.get(i));
            }
        }
    }
    
    public void clearInventory() {
        if (inventory.isEmpty()) {
            System.out.println("The inventory is already empty!");
        } else {
            inventory.clear();
            System.out.println("All items had been cleared from the inventory!");
        }
    }
}

class Shop { //Shop Methods for clients
    private List<Item> cart;
    
    public Shop() {
        this.cart = new ArrayList<>();
    }

    public void viewCatalog(Inventory inventory) {
        inventory.viewInventory();
    }

    public void addToCart(Inventory inventory) {
        Scanner input = new Scanner(System.in);
        System.out.print("Enter the PKU of the item to add to cart: ");
        String pku = input.nextLine();

        Item selectedItem = null;
        for (Item item : inventory.inventory) {
            if (item.getPKU().equals(pku)) {
                selectedItem = item;
                break;
            }
        }

        if (selectedItem != null) {
            cart.add(selectedItem);
            System.out.println("Item added to cart: " + selectedItem);
        } else {
            System.out.println("Item not found in inventory.");
        }
    }

    public void viewCart() {
        if (cart.isEmpty()) {
            System.out.println("Your cart is empty.");
        } else {
            System.out.println("Your Cart:");
            for (Item item : cart) {
                System.out.println(item);
            }
        }
    }

    public void removeFromCart() {
        Scanner input = new Scanner(System.in);
        System.out.print("Enter the PKU of the item to remove from cart: ");
        String pku = input.nextLine();

        boolean removed = cart.removeIf(item -> item.getPKU().equals(pku));
        if (removed) {
            System.out.println("Item removed from cart.");
        } else {
            System.out.println("Item not found in cart.");
        }
    }

    // Confirm purchase and show total price
    public void confirmPurchase() {
        if (cart.isEmpty()) {
            System.out.println("Your cart is empty.");
            return;
        }
        
        double total = 0;
        System.out.println("Items in your purchase:");
        for (Item item : cart) {
            System.out.println(item);
            total += Double.parseDouble(item.getPrice());
        }
        System.out.printf("Total Price: %.2f\n", total);
        
        System.out.print("Confirm purchase? (Y/N): ");
        Scanner input = new Scanner(System.in);
        String confirm = input.nextLine().toLowerCase();

        if (confirm.equals("y")) {
            System.out.println("Purchase complete, thank you!");
            cart.clear();
        } else {
            System.out.println("Purchase cancelled.");
        }
    }
}

// Where I used the concept of inheritance and composition
abstract class EmployeeUser {
    protected String username;
    
    public EmployeeUser(String username) {
        this.username = username;
    }
    public void viewInventory(Inventory inventory) {
        inventory.viewInventory();
    }
    public void findItem(Inventory inventory) {
        inventory.findItem();
    }
    
}
class Admin extends EmployeeUser {
    public Admin(String username) {
        super(username);
    }
    public void addItem(Inventory inventory) {
        inventory.addItem();
    }
    public void deleteItem(Inventory inventory) {
        inventory.deleteItem();
    }
    public void clearInventory(Inventory inventory) {
        inventory.clearInventory();
    }
}
class Operator extends EmployeeUser {
    public Operator(String username) {
        super(username);
    }
    public void addItem(Inventory inventory) {
        inventory.addItem();
    }
    public void deleteItem(Inventory inventory) {
        inventory.deleteItem();
    }
}
class Regular extends EmployeeUser {
    public Regular(String username) {
        super(username);
    }
}

File: OOPforIncompleteLab3/Mainsystem.py
class Item:
    def __init__(self, pku, name, price):
        self.pku = pku
        self.name = name
        self.price = price

    def __str__(self):
        return f"\nPKU: {self.pku}\nName: {self.name}\nPrice: {self.price}"

class Inventory:
    def __init__(self):
        self.inventory = []

    def add_item(self):
        pku = input("Enter the PKU of item: ")
        name = input("Enter the name of item: ")
        price = input("Enter the price of item: ")

        new_item = Item(pku, name, price)
        self.inventory.append(new_item)
        print("\nNew Item Added:\n", new_item)

    def delete_item(self):
        if not self.inventory:
            print("The inventory is already empty.")
            return
        pku = input("Enter the PKU of the item to delete: ")

        item_found = False
        for i, item in enumerate(self.inventory):
            if item.pku == pku:
                del self.inventory[i]
                print("The item has been deleted successfully.")
                item_found = True
                break
        if not item_found:
            print("The item does not exist.")

    def find_item(self):
        name = input("Enter the name of the item: ")
        for item in self.inventory:
            if item.name.lower() == name.lower():
                print(item)

    def view_inventory(self):
        if not self.inventory:
            print("The inventory is empty.")
        else:
            print("Inventory Catalog:")
            for item in self.inventory:
                print(item)

    def clear_inventory(self):
        if not self.inventory:
            print("The inventory is already empty!")
        else:
            self.inventory.clear()
            print("All items have been cleared from the inventory!")

class Shop:
    def __init__(self):
        self.cart = []

    def view_catalog(self, inventory):
        inventory.view_inventory()

    def add_to_cart(self, inventory):
        pku = input("Enter the PKU of the item to add to cart: ")

        selected_item = next((item for item in inventory.inventory if item.pku == pku), None)

        if selected_item:
            self.cart.append(selected_item)
            print("Item added to cart: ", selected_item)
        else:
            print("Item not found in inventory.")

    def view_cart(self):
        if not self.cart:
            print("Your cart is empty.")
        else:
            print("Your Cart:")
            for item in self.cart:
                print(item)

    def remove_from_cart(self):
        pku = input("Enter the PKU of the item to remove from cart: ")

        removed = any(item.pku == pku for item in self.cart)
        self.cart = [item for item in self.cart if item.pku != pku]

        if removed:
            print("Item removed from cart.")
        else:
            print("Item not found in cart.")

    def confirm_purchase(self):
        if not self.cart:
            print("Your cart is empty.")
            return

        total = sum(float(item.price) for item in self.cart)
        print("Items in your purchase:")
        for item in self.cart:
            print(item)
        print(f"Total Price: {total:.2f}")

        confirm = input("Confirm purchase? (Y/N): ").lower()

        if confirm == 'y':
            print("Purchase complete, thank you!")
            self.cart.clear()
        else:
            print("Purchase cancelled.")

class EmployeeUser:
    def __init__(self, username):
        self.username = username

    def view_inventory(self, inventory):
        inventory.view_inventory()

    def find_item(self, inventory):
        inventory.find_item()

class Admin(EmployeeUser):
    def add_item(self, inventory):
        inventory.add_item()

    def delete_item(self, inventory):
        inventory.delete_item()

    def clear_inventory(self, inventory):
        inventory.clear_inventory()

class Operator(EmployeeUser):
    def add_item(self, inventory):
        inventory.add_item()

    def delete_item(self, inventory):
        inventory.delete_item()

class Regular(EmployeeUser):
    pass
